daft/sql/sql.py

```diff
@@ -35,17 +39,31 @@ def sql_expr(sql: str) -> Expression:
 
 
 @PublicAPI
-def sql(sql: str, catalog: SQLCatalog) -> DataFrame:
+def sql(sql: str, catalog: Optional[SQLCatalog] = None) -> DataFrame:
     """Create a DataFrame from an SQL query.
 
     EXPERIMENTAL: This features is early in development and will change.
 
     Args:
         sql (str): SQL query to execute
+        catalog (SQLCatalog, optional): Catalog of tables to use in the query.
+            Defaults to None, in which case a catalog will be built from variables
+            in the callers scope
 
     Returns:
         DataFrame: Dataframe containing the results of the query
     """
+    if catalog is None:
+        try:
+            # Caller is back from func, analytics, annotation
+            caller_frame = inspect.currentframe().f_back.f_back.f_back  # type: ignore
+            caller_vars = {**caller_frame.f_globals, **caller_frame.f_locals}  # type: ignore
+        except AttributeError as exc:
+            # some interpreters might not implement currentframe; all reasonable
+            # errors above should be AttributeError
+            raise DaftCoreException("Cannot get caller environment, please provide a catalog") from exc
+        catalog = SQLCatalog({k: v for k, v in caller_vars.items() if isinstance(v, DataFrame)})
```

Comment:
we can probably do this in a followup PR, but it'd be nice to also support anything that can be zero copy converted into a `DataFrame`, such as: 

- anything implementing [dataframe interchange protocol](https://data-apis.org/dataframe-protocol/latest/index.html)
- [pyarrow tables](https://arrow.apache.org/docs/python/generated/pyarrow.Table.html)
- [pyarrow recordbatch](https://arrow.apache.org/docs/python/generated/pyarrow.RecordBatch.html)